instrument-starters/opentracing-spring-cloud-reactor-starter/src/main/java/io/opentracing/contrib/spring/cloud/reactor/ReactorTracingAutoConfiguration.java

```diff
@@ -37,27 +45,74 @@
  * @author Csaba Kos
  */
 @Configuration
-@AutoConfigureAfter(TracerAutoConfiguration.class)
 @ConditionalOnBean(Tracer.class)
 @ConditionalOnClass(Hooks.class)
 @ConditionalOnProperty(name = "opentracing.spring.cloud.reactor.enabled", havingValue = "true", matchIfMissing = true)
 public class ReactorTracingAutoConfiguration {
+
+  private static final Logger log = Logger.getLogger(ReactorTracingAutoConfiguration.class.getName());
+
   private static final String EXECUTOR_SERVICE_DECORATOR_KEY = ReactorTracingAutoConfiguration.class.getName();
   private static final String HOOK_KEY = ReactorTracingAutoConfiguration.class.getName();
 
-  @Autowired
-  public ReactorTracingAutoConfiguration(final Tracer tracer) {
-    Hooks.onEachOperator(HOOK_KEY, TracedSubscriber.asOperator(tracer));
-    Schedulers.setExecutorServiceDecorator(
-        EXECUTOR_SERVICE_DECORATOR_KEY,
-        (scheduler, scheduledExecutorService) ->
-            new TracedScheduledExecutorService(scheduledExecutorService, tracer)
-    );
+  @Bean
+  static HookRegisteringPostProcessor hookRegisteringPostProcessor(ObjectProvider<Tracer> tracerProvider) {
+    return new HookRegisteringPostProcessor(tracerProvider);
   }
 
-  @PreDestroy
-  public void cleanupHooks() {
-    Hooks.resetOnEachOperator(HOOK_KEY);
-    Schedulers.removeExecutorServiceDecorator(EXECUTOR_SERVICE_DECORATOR_KEY);
+  // Use a BeanDefinitionRegistryPostProcessor here to decorate Reactor as soon as possible before potential early initialization of Reactor schedulers
+  // For more detailed description, please refer to https://github.com/opentracing-contrib/java-spring-cloud/issues/214
+  private static class HookRegisteringPostProcessor implements BeanDefinitionRegistryPostProcessor {
+
+    private final ObjectProvider<Tracer> tracerProvider;
```

Comment:
I don't think that `ObjectProvider` provides any benefits over `ObjectFactory` for us. If we call `getIfAvailable()`, then we have to handle the `null` return value (which is not currently handled) and `BeansException`, and I think the handling code for both would basically be the same.

For that reason, I suggest switching to `ObjectFactory` (and `getObject()`, which will throw `BeansException` in both cases).

Reply:
I agree with what you said. I will address it.